### ISSUE TITLE
[MU4 Issue] Cannot delete or reset "auto-generated" parts

### ISSUE THREAD

**Issue by cbjeukendrup** [musescore/MuseScore#15546]
**Describe the bug**
In the Parts dialog, it is not possible to delete "auto-generated" parts, even after they have really been created by opening them. This is currently disabled on purpose, because it is not implemented; see https://github.com/musescore/MuseScore/issues/11349#issuecomment-1172462901. 

The possibility to delete an auto-generated part is necessary in the following cases:
- You messed up the part badly and want to reset it to start over again
- MuseScore messed up the part by corrupting it; that happens sometimes currently. For scores where this has happened, the best way to recover them is deleting the part(s) containing the corruption; currently that is not possible.

**Comment by sammik** [comment]
I think this needs design first.

I see several problems with #15665 behaviour.

1) In User POV, it is strange, if I delete part, and next time opening Parts dialog, "deleted" part is there again.
So I think, if deleting parts is allowed, than parts should remain deleted. May be some new button to "Recreate generic parts".
Or dont allow "deleting" generic parts and allow "recreate" or "restore default" or something like that instead.

2) If parts were not already open, "delete" is missing in generic parts. But is present in custom parts. (To reproduce, create new String quartet score, open parts dialog and try to delete part.)

**Comment by cbjeukendrup** [comment]
Yes, I agree this will need some thoughts again. @Tantacrul's idea was to hide the concept of "creating parts" from the user, because some users might be unable to understand it. That seems a good to me, as long as it doesn't have any negative consequences for users who do understand it, which it hasn't indeed. 

But when it comes to deleting parts, this design, or rather the current implementation of it, is actually only _more_ confusing: deleted parts (that were auto-created initially) suddenly reappear when you reopen the dialog, and then you can't re-delete them anymore until you open them. I'm afraid that users who don't know the underlying thoughts may experience this as janky.

So I propose that we slightly tweak the design and implementation:

1. When the user creates a new score, or when the user adds new instruments, we generate parts for the missing instruments. (Additionally, we might also do that for MS 3.x/2.x/1.x scores, which might not have all parts yet.)
2. But after that, we forget which parts were generated by MuseScore and which were user-created; all of them get equal treatment. In particular, all of them can be deleted.
3. We provide a way to easily re-create deleted parts, namely a popup inside the parts dialog with a "Create empty part" button, and a list of instruments that don't have a part so that the user can click those instruments to create a new part for them. 

Implementation detail: to save resources, we don't fully create those generated parts until the user opens or exports them. The user won't notice anything about this, except better performance.

The disadvantage is that this requires a new way to save those not-yet-fully-created parts in the mscz file, which breaks backwards compatibility. So if we make this change in 4.0.1, then 4.0.0 won't be able to open 4.0.1 scores. Not sure if that will be a bad thing in practice, but it is against the rules of semantic versioning.

As an alternative temporary solution, we could as @sammik suggests make a "reset part" button instead of a delete button. That will be sufficient to solve this issue, but not great because MuseScore forces you to have parts that you potentially don't want. 

@Tantacrul Please let me know what you prefer: 1) my suggested design tweak or 2) the alternative "reset part" solution or 3) PR https://github.com/musescore/MuseScore/pull/15665 as it is right now.

**Comment by Tantacrul** [comment]
Yeah, the limitations we ran under while implementing this required a few less than ideal solutions, which I think we can easily improve on by implementing some new designs. I definitely want to attack this problem properly, along with a few other related issues with parts, so we can finish with it once and for all. We'll discuss it during planning.

**Comment by sammik** [comment]
I like proposal by @cbjeukendrup 

Few things to have in mind.

ad 3. - recreating "generic" parts 
Should this be based by name, or by instrument?

a) If by name, than also renaming "generic part" should cause, that "generic name" appears again in "Create new part".
Example: In String quartet score, I rename "Violin I" to "Fiddle I". Than "Create new part" popup shows "Violin I". 
So (for implementation), we need list of generic names, based on list of instruments (their long names) in score.
Problem could be, if I rearange instruments in parts (I hide Violin I instrument in part and show Violin II instrument). Now I in fact miss some instrument in parts, but "Create new part" doesnt shows it in Popup.

b) If by instrument, than rule "instrument, which dont have part yet" should be restricted to "instrument, which dont have **exclusive part** yet". "Exclusive part" means, it contains only one instrument.
Example: In quartet score, I delete "Violin II" part and rename "Violin I" part to "Violins", in which I show both violins (instruments). Than "Create new part" popup should show "Violin II" and also "Violin I". 
Problem is, if I dont rename "Violin I". Than I already have "Violin I" part, but by restricted rule, it should apper in popup.
But if I use not restricted rule instead, and have "Violins" part, containing both violin instruments, they wouldnt appear in "Create new part" popup, because each of violin instruments does have part.

Conclusion - we probably need combinantion of both, by name and by instrument. Proper design I leave to the chief.

**Comment by cbjeukendrup** [comment]
@sammik Thanks!

ad a: The name for the part would be based on the current name of the instrument in the score; so if you have renamed the instrument to "Fiddle 1", then the part created for it will get the same name.

ad b: Yes, I mean "instruments that don't have an exclusive part yet" :)

--- 

@Tantacrul I understand that a bigger redesign will have to wait a bit longer, but it would be great if we could provide at least an emergency solution for MS4.0.1, so that users can fix up all those scores with corrupted parts. So what can we do?

**Comment by sammik** [comment]
@cbjeukendrup 
ad a: I meant "if I rename **part** to Fiddle I". Sorry, I have problems explain what I mean :) especcially in english.
I clarify it in my previous post.

**Comment by cbjeukendrup** [comment]
Ah, I see. Well, if there was previously a part called "Fiddle 1" for the first violin, and that part is deleted, then a newly created part for that instrument will use the instrument name rather than the name of the old part, since the old part is gone for good at the moment that the new one is created.

Regarding "name clashes" as you describe in point b: Currently there is some logic to prevent that by simply adding " 1" to the name of the newly created part. We could try to refine that logic a bit more. 

Anyway, I propose that we don't dive into all edge cases and other details yet, but first focus on something simple we can do for 4.0.1, and then sit down to compose a bigger vision for Parts.

**Comment by sammik** [comment]
> Well, if there was previously a part called "Fiddle 1" for the first violin
That was the point, how program knows, that this part is for first violin? (If logic were based by name). 

I agree for now it would be good to have something temporary working and wait for design (which seems to be much more difficualt task).

Sorry for spamming.

**Comment by cbjeukendrup** [comment]
You're by no means spamming, you have very valid points. But it's just a bit too early to start thinking about this, since we don't know yet if my proposal will make it into the final design :)

**Comment by Tantacrul** [comment]
Adding @bkunda and @oktophonie in order to specify the design.

@cbjeukendrup, given the problem you've outlined, where someone's score might be messed up in some way and they might want a nuclear option to fix it, I'm wondering if conceptually, we should frame it as a **Reset part** function, rather than describe it as **delete** function, which I think requires some lateral thinking / guesswork on the part of the user.

**Comment by cbjeukendrup** [comment]
Another idea: just create an additional "reset part" button for _every_ part (even custom ones), that will under the hood replace the part with a newly generated one containing the same instruments. This would be not massively difficult to implement.

I do think that a real "delete part" option for every part (even non-custom ones), where the part does not automatically come back, makes sense, so that we don't force the user to have parts they don't need.

**Comment by cbjeukendrup** [comment]
So, to answer your question: yes, I think that for an interim solution for 4.0.2, it makes sense to have a "reset" button for non-custom parts, and a "delete" button for custom parts, if that is the easiest for now (which it probably is).

Then for 4.1, we can make a more complete solution where every part has a fully working "reset part" _and_ "delete part" button, perhaps following my suggestion from https://github.com/musescore/MuseScore/issues/15546#issuecomment-1368498512.

**Comment by DmitryArefiev** [comment]
Retested on Win10
